### ISSUE TITLE
Support OTLP related environmental variables

### ISSUE THREAD

**Issue by Kielek** [open-telemetry/opentelemetry-dotnet#4154]
# Feature Request

Support for following environmental variables
* `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`
* `OTEL_EXPORTER_OTLP_METRICS_ENDPOINT`
*  `OTEL_EXPORTER_OTLP_LOGS_ENDPOINT`
*  `OTEL_EXPORTER_OTLP_INSECURE` `
*  `OTEL_EXPORTER_OTLP_METRICS_INSECURE`
* `OTEL_EXPORTER_OTLP_LOGS_INSECURE`
*  `OTEL_EXPORTER_OTLP_SPAN_INSECURE`
*  `OTEL_EXPORTER_OTLP_METRIC_INSECURE`
*  `OTEL_EXPORTER_OTLP_CERTIFICATE`
*  `OTEL_EXPORTER_OTLP_TRACES_CERTIFICATE`
*  `OTEL_EXPORTER_OTLP_METRICS_CERTIFICATE`
*  `OTEL_EXPORTER_OTLP_LOGS_CERTIFICATE`
*  `OTEL_EXPORTER_OTLP_CLIENT_KEY`
*  `OTEL_EXPORTER_OTLP_TRACES_CLIENT_KEY` 
* `OTEL_EXPORTER_OTLP_METRICS_CLIENT_KEY` 
* `OTEL_EXPORTER_OTLP_LOGS_CLIENT_KEY`
* `OTEL_EXPORTER_OTLP_CLIENT_CERTIFICATE`
*  `OTEL_EXPORTER_OTLP_TRACES_CLIENT_CERTIFICATE`
*  `OTEL_EXPORTER_OTLP_METRICS_CLIENT_CERTIFICATE`
*  `OTEL_EXPORTER_OTLP_LOGS_CLIENT_CERTIFICATE`
*  OTEL_EXPORTER_OTLP_COMPRESSION`
*  `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT`
*  `OTEL_EXPORTER_OTLP_METRICS_TIMEOUT`
*  `OTEL_EXPORTER_OTLP_LOGS_TIMEOUT`
*  `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL`
*  `OTEL_EXPORTER_OTLP_METRICS_PROTOCOL`
*  `OTEL_EXPORTER_OTLP_LOGS_PROTOCOL`

Requirements from https://github.com/open-telemetry/opentelemetry-specification/blob/dd47c009da1de335c540decc5d5930e46ea61916/specification/protocol/exporter.md

## Additional Context

Found while working on https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2009

**Comment by cijothomas** [comment]
1st priority:
OTEL_EXPORTER_OTLP_TRACES_PROTOCOL
OTEL_EXPORTER_OTLP_METRICS_PROTOCOL
OTEL_EXPORTER_OTLP_LOGS_PROTOCOL
OTEL_EXPORTER_OTLP_TRACES_ENDPOINT
OTEL_EXPORTER_OTLP_METRICS_ENDPOINT
OTEL_EXPORTER_OTLP_LOGS_ENDPOINT
2nd priority
OTEL_EXPORTER_OTLP_TRACES_TIMEOUT
OTEL_EXPORTER_OTLP_METRICS_TIMEOUT
OTEL_EXPORTER_OTLP_LOGS_TIMEOUT

^ From @Kielek

**Comment by pellared** [comment]
Partially duplicates https://github.com/open-telemetry/opentelemetry-dotnet/issues/2923

**Comment by CodeBlanch** [comment]
This might be really hard to do at the moment because logs/traces/metrics share a single options class. We have [talked in the past about splitting up the OtlpExporter into projects based on the transport](https://github.com/open-telemetry/opentelemetry-dotnet/issues/3421#issuecomment-1314630702). If we do that we could introduce dedicated options classes for each signal which would make this easier to achieve. It would also allow us to clean up [some hacks we use to make the single class safe to use with Options API](https://github.com/open-telemetry/opentelemetry-dotnet/blob/77cd1b03ee86e6b5dba7590081a0083fb85e91a6/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpTraceExporterHelperExtensions.cs#L83-L88).

/cc @alanwest

**Comment by Kielek** [comment]
@CodeBlanch, before we start work and split the packages for separate transport protocols support for following env. vars should be considered

* `OTEL_EXPORTER_OTLP_PROTOCOL`
* `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL`
* `OTEL_EXPORTER_OTLP_METRICS_PROTOCOL`
* `OTEL_EXPORTER_OTLP_LOGS_PROTOCOL`

Lets consider only Traces only:
We should end up with `OpenTelemetry.Exporter.OpenTelemetryProtocol.Grpc` and `OpenTelemetry.Exporter.OpenTelemetryProtocol.Http`.

I expect that we ends up with at least
```csharp
static OpenTelemetry.Trace.Grpc.OtlpTraceExporterHelperExtensions.AddGrpcOtlpExporter(this OpenTelemetry.Trace.TracerProviderBuilder! builder) -> OpenTelemetry.Trace.TracerProviderBuilder!
static OpenTelemetry.Trace.Http.OtlpTraceExporterHelperExtensions.AddHttpOtlpExporter(this OpenTelemetry.Trace.TracerProviderBuilder! builder) -> OpenTelemetry.Trace.TracerProviderBuilder!
````
How HTTP version should behave when `OTEL_EXPORTER_OTLP_PROTOCOL` or `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL` is set to grpc? Same for Grpc.
I can imagine that we have some metapackage handling this scenario, but is this expected solution?

**Comment by CodeBlanch** [comment]
I think a meta package would be fine. Or perhaps it could be something only auto-instrumentation supports? @alanwest and I have talked about having an "auto-configuration" package ([I did some exploration of that idea a while back](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3833)). Java has similar, from what I'm told.

**Comment by alanwest** [comment]
> How HTTP version should behave when OTEL_EXPORTER_OTLP_PROTOCOL or OTEL_EXPORTER_OTLP_TRACES_PROTOCOL is set to grpc?

I think a `NotSupportedException` would be appropriate in these cases.

**Comment by Kielek** [comment]
> > How HTTP version should behave when OTEL_EXPORTER_OTLP_PROTOCOL or OTEL_EXPORTER_OTLP_TRACES_PROTOCOL is set to grpc?
> 
> I think a `NotSupportedException` would be appropriate in these cases.

My understating otel-spec is that we should avoid throwing any exception influencing process start while configuring OTel.
Event-log with appropriate message will be better option (or fully ingnoring it and relay on meta-package).

**Comment by pellared** [comment]
I suggest "fully ignoring it" as even a distribution may support this env var. Getting an event-log would be misleading for the user.

**Comment by pawel-ta** [comment]
Hello, is there any update on this? Specifically for signal-specific exporter variables? It would be extremely valuable

**Comment by CodeBlanch** [comment]
@pawel-ta 

If you call the cross-cutting "UseOtlpExporter" extension (https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol#enable-otlp-exporter-for-all-signals) some of the spec signal-specific env vars are now supported!

Here is the list of currently supported keys: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpSpecConfigDefinitions.cs

**Comment by borland** [comment]
Hi @CodeBlanch - that `UseOtlpExporter` extension method looks nice, but unfortunately we are not using `Microsoft.Extensions.Hosting` to configure our OpenTelemetry; Rather we do it directly, e.g.

```csharp
var builder = OpenTelemetrySdk.CreateTracerProviderBuilder();
// ... configure builder 
builder.AddOtlpExporter();
builder.Build();
```

Using this method the default `OtlpExporterOptions` is created which ignores all the signal-specific environment variables. Is there any plan or path to fixing this in future? 

FWIW: We don't want to configure OTEL via `Microsoft.Extensions.Hosting` -- we use it to stand up an aspnetcore web server inside our application, but this happens quite late in our app's initialization; we want to get OTEL up and running as early as possible.

**Comment by wuhkuh** [comment]
Yeah, `UseOtlpExporter` is an improvement. The one thing blocking adoption for us, is the ability to disable exporters (see https://github.com/open-telemetry/opentelemetry-dotnet/issues/6864 for example).
That would make it that much more flexible.

**Comment by martincostello** [comment]
Updated the top-level description with what has been implemented and what hasn't as of v1.58.0 of the spec and v1.16.0 of the SDK.
